src/external/song.ts

```diff
@@ -17,100 +17,185 @@
+import { SongAnalysisState, SongAnalysisStates } from '../utils/constants';
+
+import { DataCenter, getDataCenter } from './dataCenterRegistry';
+
 const logger = Logger('Song');
 
-// TODO: Update kafka event parsers that use this
 export type SongAnalysis = {
-  analysisId: string;
-  analysisState: string;
-  studyId: string;
-  firstPublishedAt?: string;
-  [k: string]: any;
+	analysisId: string;
+	analysisState: string;
+	studyId: string;
+	firstPublishedAt?: string;
+	[k: string]: unknown;
+};
+
+type SongResponseAnalysesPage = {
+	analyses: SongAnalysis[];
+	totalAnalyses: number;
+	currentTotalAnalyses: number;
+};
+
+type PaginationInputs = {
+	limit: number;
+	offset: number;
+};
+
+type PaginationMetadata = PaginationInputs & {
+	responseCount: number;
+	totalCount: number;
 };
 
 export const getStudies = async (url: string) => {
-  const studiesUrl = urljoin(url, '/studies/all');
-  logger.info(`Fetching studies from: ${studiesUrl}`);
-  const res = await fetch(`${url}/studies/all`);
-  const studies = await res.json();
-  logger.info(`Retrieved ${studies.length} studies: ${studies}`);
-  return studies;
+	const studiesUrl = urljoin(url, '/studies/all');
+	logger.info(`Fetching studies from: ${studiesUrl}`);
+	const res = await fetch(`${url}/studies/all`);
+	const studies = await res.json();
+	logger.info(`Retrieved ${studies.length} studies: ${studies}`);
+	return studies;
 };
 
-/**
- * Fetch potentially very large responses from Song then return them in a batched stream.
- * @param analysesUrl
- * @returns
- */
-const fetchAnalysesInBatches = async (analysesUrl: string): Promise<Batch> => {
-  const controller = new abortController();
-  const timeoutPeriod = (await getAppConfig()).datacenter.fetchTimeout;
-  const batchSize = (await getAppConfig()).datacenter.batchSize;
-  const timeout = setTimeout(() => {
-    logger.warn(`Aborting request for analyses due to timeout after ${timeoutPeriod}`);
-    controller.abort();
-  }, timeoutPeriod);
-  try {
-    logger.info(`Fetching analyses from ${analysesUrl}`);
-    const res = await fetch(analysesUrl, {
-      signal: controller.signal,
-    });
-    const resStream = res.body;
-    const pipeline = resStream.pipe(streamArray.withParser()).pipe(new Batch({ batchSize }));
-    logger.info(`Received analyses response, returning as Batch.`);
-    return pipeline;
-  } catch (error) {
-    throw error;
-  } finally {
-    clearTimeout(timeout);
-  }
+const fetchAnalysesPage = async (inputs: {
+	dataCenter: DataCenter;
+	studyId: string;
+	offset: number;
+	limit: number;
+	analysisStates?: Set<SongAnalysisState>;
+}): Promise<{ data: SongAnalysis[] } & PaginationMetadata> => {
+	try {
+		const {
+			dataCenter: { centerId, songUrl },
+			studyId,
+			limit,
+			offset,
+			analysisStates,
+		} = inputs;
+		const analysisStatesQueryValue = analysisStates
+			? Array.from(analysisStates).join(',')
+			: SongAnalysisStates.PUBLISHED || SongAnalysisStates.PUBLISHED;
```

Comment:
`SongAnalysisStates.PUBLISHED || SongAnalysisStates.PUBLISHED;` - unclear what this is doing - maybe no need for the OR?

Reply:
Thanks for catching this, this doesn't do what I wanted.

The intention is to make sure that there is always at least one analysis state presented, so an empty Set would still apply PUBLISHED.

Reply:
Does it need to be:
```js
 (analysisStates? Array.from(analysisStates).join(',') : SongAnalysisStates.PUBLISHED ) || SongAnalysisStates.PUBLISHED;
```
in that case? else the OR gets evaluated first , then the ternary

Reply:
Yeah I think that's what my brain was attempting. I've updated it instead to:

```ts
const analysisStatesQueryValue =
			analysisStates && analysisStates.size ? Array.from(analysisStates).join(',') : SongAnalysisStates.PUBLISHED;
```